Pass `@warn` as warnings to webpack

At the moment I'm working on a project with a workflow that uses `sass-loader` in order to compile SCSS.

There is a Sass mixin with a `@warn` directive in it. At the moment it outputs warnings such as this to `stdout`:
```
WARNING: Warning message
         on line 56 of src/test.scss, in function `something`
```

I think it could be more useful if Webpack was aware of the warning. In the same way that `eslint-loader` can raise warnings. e.g.:

```
/path/to/code/file.js
  1:20  warning  'foobar' is defined but never used  no-unused-vars
```

Is this something I can do right now with `sass-loader` in any way? I think developers using similar workflows could benefit from seeing the warnings in a more organized fashion coming out of Webpack, rather than scattered warnings straight into `stdout`

--- a/README.md
+++ b/README.md
@@ -119,13 +119,14 @@ Thankfully there are a two solutions to this problem:
 
 ## Options
 
-|                   Name                    |         Type         |                 Default                 | Description                                                       |
-| :---------------------------------------: | :------------------: | :-------------------------------------: | :---------------------------------------------------------------- |
-|  **[`implementation`](#implementation)**  |  `{Object\|String}`  |                 `sass`                  | Setup Sass implementation to use.                                 |
-|     **[`sassOptions`](#sassoptions)**     | `{Object\|Function}` | defaults values for Sass implementation | Options for Sass.                                                 |
-|       **[`sourceMap`](#sourcemap)**       |     `{Boolean}`      |           `compiler.devtool`            | Enables/Disables generation of source maps.                       |
-|  **[`additionalData`](#additionaldata)**  | `{String\|Function}` |               `undefined`               | Prepends/Appends `Sass`/`SCSS` code before the actual entry file. |
-| **[`webpackImporter`](#webpackimporter)** |     `{Boolean}`      |                 `true`                  | Enables/Disables the default Webpack importer.                    |
+|                     Name                      |         Type         |                 Default                 | Description                                                       |
+| :-------------------------------------------: | :------------------: | :-------------------------------------: | :---------------------------------------------------------------- |
+|    **[`implementation`](#implementation)**    |  `{Object\|String}`  |                 `sass`                  | Setup Sass implementation to use.                                 |
+|       **[`sassOptions`](#sassoptions)**       | `{Object\|Function}` | defaults values for Sass implementation | Options for Sass.                                                 |
+|         **[`sourceMap`](#sourcemap)**         |     `{Boolean}`      |           `compiler.devtool`            | Enables/Disables generation of source maps.                       |
+|    **[`additionalData`](#additionaldata)**    | `{String\|Function}` |               `undefined`               | Prepends/Appends `Sass`/`SCSS` code before the actual entry file. |
+|   **[`webpackImporter`](#webpackimporter)**   |     `{Boolean}`      |                 `true`                  | Enables/Disables the default Webpack importer.                    |
+| **[`warnRuleAsWarning`](#warnruleaswarning)** |     `{Boolean}`      |                 `false`                 | Treats the `@warn` rule as a webpack warning.                     |
 
 ### `implementation`
 
@@ -595,6 +596,65 @@ module.exports = {
             loader: "sass-loader",
             options: {
               webpackImporter: false,
+            },
+          },
+        ],
+      },
+    ],
+  },
+};
+```
+
+### `warnRuleAsWarning`
+
+Type: `Boolean`
+Default: `false`
+
+Treats the `@warn` rule as a webpack warning.
+
+> ℹ️ It will be `true` by default in the next major release.
+
+**style.scss**
+
+```scss
+$known-prefixes: webkit, moz, ms, o;
+
+@mixin prefix($property, $value, $prefixes) {
+  @each $prefix in $prefixes {
+    @if not index($known-prefixes, $prefix) {
+      @warn "Unknown prefix #{$prefix}.";
+    }
+
+    -#{$prefix}-#{$property}: $value;
+  }
+  #{$property}: $value;
+}
+
+.tilt {
+  // Oops, we typo'd "webkit" as "wekbit"!
+  @include prefix(transform, rotate(15deg), wekbit ms);
+}
+```
+
+The presented code will throw webpack warning instead logging.
+
+To ignore unnecessary warnings you can use the [ignoreWarnings](https://webpack.js.org/configuration/other-options/#ignorewarnings) option.
+
+**webpack.config.js**
+
+```js
+module.exports = {
+  module: {
+    rules: [
+      {
+        test: /\.s[ac]ss$/i,
+        use: [
+          "style-loader",
+          "css-loader",
+          {
+            loader: "sass-loader",
+            options: {
+              warnRuleAsWarning: true,
             },
           },
         ],

--- a/src/SassError.js
+++ b/src/SassError.js
@@ -3,6 +3,7 @@ class SassError extends Error {
     super();
 
     this.name = "SassError";
+    // TODO remove me in the next major release
     this.originalSassError = sassError;
     this.loc = {
       line: sassError.line,

--- a/src/SassWarning.js
+++ b/src/SassWarning.js
@@ -1,0 +1,17 @@
+class SassWarning extends Error {
+  constructor(warning, options) {
+    super(warning);
+
+    this.name = "SassWarning";
+    this.hideStack = true;
+
+    if (options.span) {
+      this.loc = {
+        line: options.span.start.line,
+        column: options.span.start.column,
+      };
+    }
+  }
+}
+
+export default SassWarning;

--- a/src/options.json
+++ b/src/options.json
@@ -48,6 +48,11 @@
       "description": "Enables/Disables default `webpack` importer.",
       "link": "https://github.com/webpack-contrib/sass-loader#webpackimporter",
       "type": "boolean"
+    },
+    "warnRuleAsWarning": {
+      "description": "Treats the '@warn' rule as a webpack warning.",
+      "link": "https://github.com/webpack-contrib/sass-loader#warnruleaswarning",
+      "type": "boolean"
     }
   },
   "additionalProperties": false

--- a/src/utils.js
+++ b/src/utils.js
@@ -4,6 +4,8 @@ import path from "path";
 import { klona } from "klona/full";
 import async from "neo-async";
 
+import SassWarning from "./SassWarning";
+
 function getDefaultSassImplementation() {
   let sassImplPkg = "sass";
 
@@ -229,6 +231,9 @@ async function getSassOptions(
   }
 
   if (!options.logger) {
+    // TODO set me to `true` by default in the next major release
+    const needEmitWarning = loaderOptions.warnRuleAsWarning === true;
+
     const logger = loaderContext.getLogger("sass-loader");
     const formatSpan = (span) =>
       `${span.url || "-"}:${span.start.line}:${span.start.column}: `;
@@ -262,7 +267,13 @@ async function getSassOptions(
           builtMessage += `\n\n${loggerOptions.stack}`;
         }
 
-        logger.warn(builtMessage);
+        if (needEmitWarning) {
+          loaderContext.emitWarning(
+            new SassWarning(builtMessage, loggerOptions)
+          );
+        } else {
+          logger.warn(builtMessage);
+        }
       },
     };
   }

--- a/test/validate-options.test.js
+++ b/test/validate-options.test.js
@@ -52,6 +52,10 @@ describe("validate options", () => {
       success: [true, false],
       failure: ["string"],
     },
+    warnRuleAsWarning: {
+      success: [true, false],
+      failure: ["string"],
+    },
     unknown: {
       success: [],
       failure: [1, true, false, "test", /test/, [], {}, { foo: "bar" }],

--- a/test/warnRuleAsWarning.test.js
+++ b/test/warnRuleAsWarning.test.js
@@ -1,0 +1,151 @@
+import url from "url";
+
+import dartSass from "sass";
+
+import { isSupportedFibers } from "../src/utils";
+
+import {
+  compile,
+  getCodeFromBundle,
+  getCodeFromSass,
+  getCompiler,
+  getErrors,
+  getImplementationByName,
+  getTestId,
+  getWarnings,
+} from "./helpers";
+
+jest.setTimeout(60000);
+
+let Fiber;
+const implementations = [dartSass];
+const syntaxStyles = ["scss", "sass"];
+
+describe("loader", () => {
+  beforeAll(async () => {
+    if (isSupportedFibers()) {
+      const { default: fibers } = await import("fibers");
+      Fiber = fibers;
+    }
+  });
+
+  beforeEach(() => {
+    if (isSupportedFibers()) {
+      // The `sass` (`Dart Sass`) package modify the `Function` prototype, but the `jest` lose a prototype
+      Object.setPrototypeOf(Fiber, Function.prototype);
+    }
+  });
+
+  implementations.forEach((implementation) => {
+    const [implementationName] = implementation.info.split("\t");
+
+    syntaxStyles.forEach((syntax) => {
+      it(`should not emit warning by default (${implementationName}) (${syntax})`, async () => {
+        const testId = getTestId("logging", syntax);
+        const options = {
+          implementation: getImplementationByName(implementationName),
+        };
+        const compiler = getCompiler(testId, { loader: { options } });
+        const stats = await compile(compiler);
+        const codeFromBundle = getCodeFromBundle(stats, compiler);
+        const codeFromSass = getCodeFromSass(testId, options);
+        const logs = [];
+
+        for (const [name, value] of stats.compilation.logging) {
+          if (/sass-loader/.test(name)) {
+            logs.push(
+              value.map((item) => {
+                return {
+                  type: item.type,
+                  args: item.args.map((arg) =>
+                    arg
+                      .replace(url.pathToFileURL(__dirname), "file:///<cwd>")
+                      .replace(/\\/g, "/")
+                  ),
+                };
+              })
+            );
+          }
+        }
+
+        expect(codeFromBundle.css).toBe(codeFromSass.css);
+        expect(codeFromBundle.css).toMatchSnapshot("css");
+        expect(getWarnings(stats)).toMatchSnapshot("warnings");
+        expect(getErrors(stats)).toMatchSnapshot("errors");
+        expect(logs).toMatchSnapshot("logs");
+      });
+
+      it(`should not emit warning when 'false' used (${implementationName}) (${syntax})`, async () => {
+        const testId = getTestId("logging", syntax);
+        const options = {
+          implementation: getImplementationByName(implementationName),
+          warnRuleAsWarning: false,
+        };
+        const compiler = getCompiler(testId, { loader: { options } });
+        const stats = await compile(compiler);
+        const codeFromBundle = getCodeFromBundle(stats, compiler);
+        const codeFromSass = getCodeFromSass(testId, options);
+        const logs = [];
+
+        for (const [name, value] of stats.compilation.logging) {
+          if (/sass-loader/.test(name)) {
+            logs.push(
+              value.map((item) => {
+                return {
+                  type: item.type,
+                  args: item.args.map((arg) =>
+                    arg
+                      .replace(url.pathToFileURL(__dirname), "file:///<cwd>")
+                      .replace(/\\/g, "/")
+                  ),
+                };
+              })
+            );
+          }
+        }
+
+        expect(codeFromBundle.css).toBe(codeFromSass.css);
+        expect(codeFromBundle.css).toMatchSnapshot("css");
+        expect(getWarnings(stats)).toMatchSnapshot("warnings");
+        expect(getErrors(stats)).toMatchSnapshot("errors");
+        expect(logs).toMatchSnapshot("logs");
+      });
+
+      it(`should not emit warning when 'true' used (${implementationName}) (${syntax})`, async () => {
+        const testId = getTestId("logging", syntax);
+        const options = {
+          implementation: getImplementationByName(implementationName),
+          warnRuleAsWarning: true,
+        };
+        const compiler = getCompiler(testId, { loader: { options } });
+        const stats = await compile(compiler);
+        const codeFromBundle = getCodeFromBundle(stats, compiler);
+        const codeFromSass = getCodeFromSass(testId, options);
+        const logs = [];
+
+        for (const [name, value] of stats.compilation.logging) {
+          if (/sass-loader/.test(name)) {
+            logs.push(
+              value.map((item) => {
+                return {
+                  type: item.type,
+                  args: item.args.map((arg) =>
+                    arg
+                      .replace(url.pathToFileURL(__dirname), "file:///<cwd>")
+                      .replace(/\\/g, "/")
+                  ),
+                };
+              })
+            );
+          }
+        }
+
+        expect(codeFromBundle.css).toBe(codeFromSass.css);
+        expect(codeFromBundle.css).toMatchSnapshot("css");
+        expect(getWarnings(stats)).toMatchSnapshot("warnings");
+        expect(getErrors(stats)).toMatchSnapshot("errors");
+        expect(logs).toMatchSnapshot("logs");
+      });
+    });
+  });
+});

